### PR TITLE
Fix panic on malformed linked event chains

### PR DESCRIPTION
**Pull request by batiati**
A malformed message with the last event with `flags.linked == true` causes TigerBeetle to panic on [this assert](https://github.com/tigerbeetledb/tigerbeetle/blob/bf4f541f83b93bbf50070e33d47e41078893b2df/src/state_machine.zig#L453), documented with a `TODO`:

```zig
// TODO client.zig: Validate that batch chains are always well-formed and closed.
// This is programming error and we should raise an exception for this in the client ASAP.
assert(chain == null);
```

The comment suggested handling malformed messages at the client level, however, this PR follows a different approach since we might not have control over all client implementations.

- Introduced a new `malformed_linked_chain` result;
I'm not sure about the impact of this new enum on existing clients, but I'm assuming since we are still beta, they can be quickly updated.

- All events in a malformed linked chain fail together.
Other chains and individual events in the same batch can execute independently of the malformed one.

This PR includes tests asserting this behavior.

### CODE REVIEW

**Comment by batiati** [comment]
cc @sentientwaffle

File: src/tigerbeetle.zig [hunk 1]
```diff
@@ -120,6 +120,8 @@ pub const CreateAccountResult = enum(u32) {
     exists_with_different_credits_pending,
     exists_with_different_credits_posted,
     exists,
+
+    malformed_linked_chain,
```
Comment (sentientwaffle) [inline src/tigerbeetle.zig]: Move the error up near `linked_event_failed`. Rename it to have a `"linked_"` prefix (e.g. `linked_chain_open`).

Also, new error codes require updates to `src/c/tb_client.h`.

File: src/state_machine.zig [hunk 2]
```diff
@@ -403,17 +403,51 @@ pub fn StateMachineType(comptime Storage: type) type {
 
             var chain: ?usize = null;
```
Comment (sentientwaffle) [inline src/state_machine.zig]: For consistency, let's rely on the existing chain rollback mechanism. That way, only the last event in the chain has the specific error, and all others are `linked_event_failed`.

This diff is based on the original state machine:

```diff
--- a/src/state_machine.zig
+++ b/src/state_machine.zig
@@ -409,11 +409,13 @@ pub fn StateMachineType(comptime Storage: type) type {
                     chain = index;
                     assert(chain_broken == false);
                 }
-                const result = if (chain_broken) .linked_event_failed else switch (operation) {
-                    .create_accounts => self.create_account(event),
-                    .create_transfers => self.create_transfer(event),
-                    else => unreachable,
-                };
+                const result = if (chain_broken) .linked_event_failed
+                    else if (event.flags.linked and index == events.len - 1) .malformed_linked_chain
+                    else switch (operation) {
+                        .create_accounts => self.create_account(event),
+                        .create_transfers => self.create_transfer(event),
+                        else => unreachable,
+                    };
                 log.debug("{s} {}/{}: {}: {}", .{
                     @tagName(operation),
                     index + 1,
@@ -448,10 +450,6 @@ pub fn StateMachineType(comptime Storage: type) type {
                     chain_broken = false;
                 }
             }
-            // TODO client.zig: Validate that batch chains are always well-formed and closed.
-            // This is programming error and we should raise an exception for this in the client ASAP.
-            assert(chain == null);
-            assert(chain_broken == false);

             return @sizeOf(Result(operation)) * count;
         }
```

File: src/state_machine.zig [hunk 3]
```diff
@@ -1551,6 +1584,78 @@ test "linked accounts" {
     // All our rollback handlers appear to be commutative.
 }
 
+test "malformed linked accounts" {
+    var accounts = [_]Account{
+
```
Comment (sentientwaffle) [inline src/state_machine.zig]:
```suggestion
```

File: src/state_machine.zig [hunk 4]
```diff
@@ -1551,6 +1584,78 @@ test "linked accounts" {
...
+        &[_]CreateAccountsResult{
+            .{ .index = 3, .result = .malformed_linked_chain },
+            .{ .index = 4, .result = .malformed_linked_chain },
+        },
+        results,
+    );
+
+    try expectEqual(accounts[0], state_machine.get_account(accounts[0].id).?.*);
+    try expectEqual(accounts[1], state_machine.get_account(accounts[1].id).?.*);
+    try expectEqual(accounts[2], state_machine.get_account(accounts[2].id).?.*);
```
Comment (sentientwaffle) [inline src/state_machine.zig:1597]: Verify that accounts 3+4 are missing.

File: src/state_machine.zig [hunk 5]
```diff
@@ -1551,6 +1584,78 @@ test "linked accounts" {
...
+    const size = state_machine.commit(0, 0, .create_accounts, input, output);
+    const results = mem.bytesAsSlice(CreateAccountsResult, output[0..size]);
+
+    try expectEqualSlices(
+        CreateAccountsResult,
+        &[_]CreateAccountsResult{
+            .{ .index = 0, .result = .malformed_linked_chain },
+        },
+        results,
+    );
```
Comment (sentientwaffle) [inline src/state_machine.zig:1672]: Verify that the account does not exist.

**Review by sentientwaffle — CHANGES_REQUESTED** [review]

File: src/state_machine.zig (hunk 2 above)
Comment (batiati) [reply src/state_machine.zig]: @sentientwaffle, by "rely on the existing chain rollback mechanism"

Did you mean letting the whole chain execute normally and then rollbacking at the end when we detected the `chain != null`?
I had started doing it that way, but I assumed this rollback could be expensive and better if avoided.

Comment (sentientwaffle) [reply src/state_machine.zig]: > Did you mean letting the whole chain execute normally and then rollbacking at the end when we detected the `chain != null`?

Yes, rolling back to the start of the (invalid) chain.

>  I had started doing it that way, but I assumed this rollback could be expensive and better if avoided.

This case is rare — it is only hit when the client is used wrong (i.e. a bug in the application code). So the performance of the rollback isn't an issue.

**Review by sentientwaffle — DISMISSED** [review]
LGTM. @jorangreef any suggestions on the error name, or is `linked_chain_open` good?

File: src/tigerbeetle.zig (hunk 1 above)
Comment (jorangreef) [reply src/tigerbeetle.zig]: `linked_chain_open` is exactly what I was going to suggest as a variable name elsewhere (`chain_open`). Nice.

Comment (jorangreef) [reply src/tigerbeetle.zig]: We'll need to update the Node, Go and Java clients for this too.

File: src/state_machine.zig (hunk 2 above)
Comment (jorangreef) [reply src/state_machine.zig]: There is a separate performance cost that comes in, which is the additional branch on every event, to test flags and index. I don't know if we can ride on another condition to get rid of this also? For 8k events in a batch, with such a hot loop, this kind of branching is significant. The less conditions in the data plane here the better.

**Comment by jorangreef** [comment]
Let's go with `linked_event_chain_open` to share the `linked_event` prefix, and since that is also the name of the feature (“Linked Events”).

File: src/state_machine.zig [hunk 6]
```diff
@@ -1551,6 +1551,82 @@ test "linked accounts" {
     // All our rollback handlers appear to be commutative.
 }
 
+test "malformed linked accounts" {
+    var accounts = [_]Account{
+        // A chain of 3 events (the last event in the chain closes the chain with linked=false)
```
Comment (jorangreef) [inline src/state_machine.zig]:
```suggestion
        // A chain of 3 events (the last event in the chain closes the chain with linked=false):
```

We use a colon since it refers to a block of lines below. Otherwise we would have a full stop.

File: src/state_machine.zig [hunk 7]
```diff
@@ -1551,6 +1551,82 @@ test "linked accounts" {
...
+        mem.zeroInit(Account, .{ .id = 1, .code = 1, .ledger = 1, .flags = .{ .linked = true } }),
+        mem.zeroInit(Account, .{ .id = 2, .code = 1, .ledger = 1, .flags = .{ .linked = true } }),
+        mem.zeroInit(Account, .{ .id = 3, .code = 1, .ledger = 1 }),
+
+        // An open chain of 2 events
```
Comment (jorangreef) [inline src/state_machine.zig]:
```suggestion
        // An open chain of 2 events:
```

File: src/state_machine.zig [hunk 8]
```diff
@@ -409,7 +409,11 @@ pub fn StateMachineType(comptime Storage: type) type {
                     chain = index;
                     assert(chain_broken == false);
                 }
-                const result = if (chain_broken) .linked_event_failed else switch (operation) {
+                const result = if (chain_broken)
+                    .linked_event_failed
+                else if (event.flags.linked and index == events.len - 1)
```
Comment (jorangreef) [inline src/state_machine.zig]: We could hide this redundant branch on `event.flags.linked` for performance by refactoring this into the existing `if (event.flags.linked...` branch that we already have on line 408, doing this by means of a `blk` when assigning to `result`.

Comment (jorangreef) [reply src/state_machine.zig]: Otherwise, we're adding an extra branch for every event processed by TigerBeetle, even where we're only checking the last event, and only if it's linked. This is also random runtime data so it's hard to predict.

Comment (jorangreef) [reply src/state_machine.zig]: Since we already also have the branch that checks for whether an event is linked, we can just fold this cost into that.

File: src/state_machine.zig [hunk 9]
```diff
@@ -409,7 +409,11 @@ pub fn StateMachineType(comptime Storage: type) type {
...
+                    .linked_chain_open
```
Comment (jorangreef) [inline src/state_machine.zig]:
```suggestion
                    .linked_event_chain_open
```

File: src/state_machine.zig [hunk 10]
```diff
@@ -448,10 +452,6 @@ pub fn StateMachineType(comptime Storage: type) type {
                     chain_broken = false;
                 }
             }
-            // TODO client.zig: Validate that batch chains are always well-formed and closed.
-            // This is programming error and we should raise an exception for this in the client ASAP.
-            assert(chain == null);
-            assert(chain_broken == false);
```
Comment (jorangreef) [inline src/state_machine.zig]: We should leave these in, since now they should never be hit, unless we have a bug in our linked event code above, or unless we're not cleaning up for the `linked_event_chain_open` result.

File: src/state_machine.zig [hunk 11]
```diff
@@ -1551,6 +1551,82 @@ test "linked accounts" {
...
+
+    try expectEqual(accounts[0], state_machine.get_account(accounts[0].id).?.*);
+    try expectEqual(accounts[1], state_machine.get_account(accounts[1].id).?.*);
+    try expectEqual(accounts[2], state_machine.get_account(accounts[2].id).?.*);
+
+    try expectEqual(@as(?*const Account, null), state_machine.get_account(accounts[3].id));
+    try expectEqual(@as(?*const Account, null), state_machine.get_account(accounts[4].id));
+}
+
+test "single linked account batch" {
```
Comment (jorangreef) [inline src/state_machine.zig]:
```suggestion
test "linked_event_chain_open for a batch of 1" {
```

File: src/state_machine.zig [hunk 12]
```diff
@@ -1551,6 +1551,82 @@ test "linked accounts" {
     // All our rollback handlers appear to be commutative.
 }
 
+test "malformed linked accounts" {
```
Comment (jorangreef) [inline src/state_machine.zig]:
```suggestion
test "linked_event_chain_open" {
```

File: src/state_machine.zig (hunk 11 above)
Comment (jorangreef) [reply src/state_machine.zig]: Nice test case to have by the way, since it explores the boundary condition.

**Review by jorangreef — CHANGES_REQUESTED** [review]
Thanks @batiati awesome work! We can still just try and hide the extra branch mispredict, by folding it into an existing condition.

File: src/state_machine.zig [hunk 13]
```diff
@@ -437,19 +444,19 @@ pub fn StateMachineType(comptime Storage: type) type {
                                 count += 1;
                             }
                         } else {
-                            assert(result == .linked_event_failed);
+                            assert(result == .linked_event_failed or result == .linked_event_chain_open);
```
Comment (batiati) [inline src/state_machine.zig:447]: I added a new test "linked_event_chain_open for an already failed batch" that would trip this assert without the `or`.

File: src/state_machine.zig [hunk 14]
```diff
@@ -437,19 +444,19 @@ pub fn StateMachineType(comptime Storage: type) type {
...
                         }
                     }
                     results[count] = .{ .index = @intCast(u32, index), .result = result };
                     count += 1;
                 }
-                if (!event.flags.linked and chain != null) {
+                if ((!event.flags.linked and chain != null) or
+                    result == .linked_event_chain_open)
+                {
```
Comment (batiati) [inline src/state_machine.zig]: I had to add an `or` here to keep the assertion `assert(chain == null)` at the end.
Not sure about that.

**Comment by sentientwaffle** [comment]
@batiati please can you rebase your branch on `origin/main` so that CI passes? (We were reworking it earlier, but that should fix it). I would do it but the PR is from your fork.

**Review by sentientwaffle — DISMISSED** [review]
LGTM. Thank you!

I was a little surprised by the `linked_event_chain_open for an already failed batch` test case, where a batch can have both `linked_event_chain_open` and another error (e.g. `exists_with_different_flags`), but it makes sense in retrospect. The "exists" error occurs first chronologically, but the `linked_event_chain_open` still needs to be included because it takes "precedence", i.e. it is a lower-level client error.

**Comment by sentientwaffle** [comment]
@jorangreef it looks like because you have a "change requested" review unresolved it won't allow me to squash + merge until you approve :wink:.

File: src/state_machine.zig [hunk 15]
```diff
@@ -405,14 +405,21 @@ pub fn StateMachineType(comptime Storage: type) type {
...
-                    .create_transfers => self.create_transfer(event),
-                    else => unreachable,
+                const result = blk: {
+                    if (event.flags.linked) {
+                        if (index == events.len - 1) {
+                            break :blk .linked_event_chain_open;
+                        } else if (chain == null) {
+                            chain = index;
+                            assert(chain_broken == false);
+                        }
```
Comment (jorangreef) [inline src/state_machine.zig]: I think we may want to have these as concurrent clauses?

i.e. First open the chain if null, because that doesn't break, and then break if necessary for an open chain.

This keeps the logic the same as how you were otherwise going to do the switch down below.

File: src/state_machine.zig (hunk 14 above)
Comment (jorangreef) [reply src/state_machine.zig]: I don't know if my new comment up top will help with this?

**Review by jorangreef — DISMISSED** [review]
Approving to work around GitHub so that @sentientwaffle can squash... but would be great to get those clauses concurrent first.

File: src/state_machine.zig (hunk 14 above)
Comment (batiati) [reply src/state_machine.zig]: I was referring to the new branch introduced by the `or` condition, after all, we refactored to remove a branch, but yet we introduced another one 🤔.

But I think I got it right, I just refactored to `if (chain != null and (!event.flags.linked or result == .linked_event_chain_open))`, so we fold the second branch under the `chain != null`

**Review by jorangreef — APPROVED** [review]
Looks great @batiati ! 🔥
